File: backend/app/priority_flags.py
"""
Low Hanging and Top 10 — the two capped, ranked Opportunity picks.

Mirrors frontend/src/lib/priorityFlags.ts exactly, on purpose: this is
Opportunity business logic (a 5/10 cap, mutual exclusivity, a rank the user
picks rather than one the system assigns), not plumbing, so both sides agree
on the same rule rather than one trusting the other. The frontend still
enforces it too — this module is what makes that enforcement real once the
request no longer passes through MSW.
"""

from dataclasses import dataclass
from typing import Mapping, Sequence


@dataclass(frozen=True)
class PriorityFlag:
    flag: str
    rank: str
    cap: int
    label: str


PRIORITY_FLAGS: tuple[PriorityFlag, ...] = (
    PriorityFlag("is_low_hanging", "low_hanging_rank", 5, "Low Hanging"),
    PriorityFlag("is_top_10", "top_10_rank", 10, "Top 10"),
)


class PriorityFlagError(ValueError):
    """Raised with a user-facing message — the router turns this into a 409."""


def resolve_priority_flag_patch(
    all_opportunities: Sequence[Mapping[str, object]],
    current_id: str,
    current: Mapping[str, object] | None,
    patch: Mapping[str, object],
) -> dict[str, object]:
    """
    Validates and enriches one write against the priority-flag rules.

    Only looks at flags actually present in `patch` — a save that never
    touches is_low_hanging or is_top_10 comes back unchanged. The chosen rank
    must already be in `patch` too: this never assigns one, only checks it.

    Returns the patch with is_low_hanging/is_top_10's rank fields filled in
    (on a real pick) or cleared (on an uncheck), and the OTHER flag/rank pair
    cleared when one is freshly picked, since the two are mutually exclusive.
    Raises PriorityFlagError, with a message safe to show the user, when the
    requested rank is missing, out of range, or already held by another row.
    """
    out: dict[str, object] = dict(patch)

    for pf in PRIORITY_FLAGS:
        if pf.flag not in patch:
            continue

        turning_on = bool(patch[pf.flag])
        was_on = bool(current.get(pf.flag)) if current is not None else False

        if turning_on:
            requested = patch.get(pf.rank)
            if not isinstance(requested, int) or isinstance(requested, bool):
                raise PriorityFlagError(f"Pick a rank between 1 and {pf.cap} for {pf.label}.")
            if not (1 <= requested <= pf.cap):
                raise PriorityFlagError(f"Pick a rank between 1 and {pf.cap} for {pf.label}.")

            taken = {
                row.get(pf.rank)
                for row in all_opportunities
                if row.get("id") != current_id and row.get(pf.flag)
            }
            if requested in taken:
                raise PriorityFlagError(
                    f"{pf.label} rank {requested} is already taken — pick another open rank."
                )
            out[pf.rank] = requested

            # Mutually exclusive: picking one clears the other.
            for other in PRIORITY_FLAGS:
                if other.flag == pf.flag:
                    continue
                out[other.flag] = False
                out[other.rank] = None
        elif not turning_on and was_on:
            out[pf.rank] = None

    return out

File: frontend/src/lib/priorityFlags.ts
/**
 * Low Hanging and Top 10 — the two capped, ranked opportunity picks.
 *
 * Single source of truth for the cap sizes on this side — the server enforces
 * them (app/priority_flags.py) — used by the rank-picker control that only ever offers a currently
 * open rank in the first place. Rank is always chosen by the user — never
 * assigned silently — so the handler's job is to validate the choice, not to
 * make it.
 */

type Item = Record<string, unknown>

export interface PriorityFlag {
  flag: string
  rank: string
  cap: number
  label: string
}

export const PRIORITY_FLAGS: readonly PriorityFlag[] = [
  { flag: 'is_low_hanging', rank: 'low_hanging_rank', cap: 5, label: 'Low Hanging' },
  { flag: 'is_top_10', rank: 'top_10_rank', cap: 10, label: 'Top 10' },
]

/** The pick a record holds, and its rank — see priorityFlagOf. */
export interface HeldPriorityFlag {
  pick: PriorityFlag
  /** Null when the flag is on but no rank was stored (legacy rows). */
  rank: number | null
}

/**
 * Which priority pick a record holds, if any.
 *
 * At most one: the two are mutually exclusive, enforced here in
 * resolvePriorityFlagPatch and on the server. A row that somehow carries both
 * — only possible if data was written around both — reports the first in
 * PRIORITY_FLAGS order rather than drawing two marks for a state the rules
 * say cannot exist.
 */
export function priorityFlagOf(row: Item | undefined): HeldPriorityFlag | null {
  if (!row) return null
  for (const pick of PRIORITY_FLAGS) {
    if (!row[pick.flag]) continue
    const n = Number(row[pick.rank])
    return { pick, rank: Number.isFinite(n) && n > 0 ? n : null }
  }
  return null
}

/** Ranks 1..cap already held by some OTHER record with this flag on. */
export function takenRanksFor(
  allOpportunities: Item[],
  currentId: string,
  flag: string,
  rank: string
): Set<number> {
  const taken = new Set<number>()
  for (const r of allOpportunities) {
    if (r.id === currentId || !r[flag]) continue
    const n = Number(r[rank])
    if (Number.isFinite(n)) taken.add(n)
  }
  return taken
}

/**
 * Resolves one PUT patch against the opportunities collection's priority-flag
 * rules: the chosen rank must be a real, currently open slot, and picking one
 * flag clears the other (Low Hanging and Top 10 are mutually exclusive).
 *
 * Only looks at the flags actually present in `patch` — a save that never
 * touches is_low_hanging or is_top_10 passes through untouched.
 */
export function resolvePriorityFlagPatch(
  allOpportunities: Item[],
  currentId: string,
  current: Item | undefined,
  patch: Item
): { ok: true; patch: Item } | { ok: false; message: string } {
  const out = { ...patch }

  for (const { flag, rank, cap, label } of PRIORITY_FLAGS) {
    if (!(flag in patch)) continue

    const turningOn = Boolean(patch[flag])
    const wasOn = Boolean(current?.[flag])

    if (turningOn) {
      const requested = Number(patch[rank])
      if (!Number.isInteger(requested) || requested < 1 || requested > cap) {
        return { ok: false, message: `Pick a rank between 1 and ${cap} for ${label}.` }
      }

      const taken = takenRanksFor(allOpportunities, currentId, flag, rank)
      if (taken.has(requested)) {
        return {
          ok: false,
          message: `${label} rank ${requested} is already taken — pick another open rank.`,
        }
      }
      out[rank] = requested

      // Mutually exclusive: picking one clears the other.
      for (const other of PRIORITY_FLAGS) {
        if (other.flag === flag) continue
        out[other.flag] = false
        out[other.rank] = null
      }
    } else if (!turningOn && wasOn) {
      out[rank] = null
    }
  }

  return { ok: true, patch: out }
}
